chain/client/src/sync/block.rs

```diff
@@ -181,39 +181,58 @@ impl BlockSync {
         // the rest of the logic succeeds.
         // TODO: If this code fails we should retry ASAP. Shouldn't we?
         let chain_head = chain.head()?;
+        let header_head = chain.header_head()?;
+        let _span = tracing::debug_span!(
+            target: "sync",
+            "block_sync",
+            head_last_block_hash = ?chain_head.last_block_hash,
+            head_height = chain_head.height,
+            header_head_height = header_head.height,
+        )
+        .entered();
+
         self.last_request =
             Some(BlockSyncRequest { head: chain_head.last_block_hash, when: self.clock.now_utc() });
 
         // The last block on the canonical chain that is processed (is in store).
         let reference_hash = self.get_last_processed_block(chain)?;
 
+        // Assume that peers are configured to keep as many epochs does this
+        // node and expect peers to have blocks in the range
+        // [gc_stop_height, header_head.last_block_hash].
+        let gc_stop_height = chain.runtime_adapter.get_gc_stop_height(&header_head.last_block_hash);
+
         // Look ahead for max_block_requests block headers and add requests for
         // blocks that we don't have yet.
-        let mut requests = vec![];
         let mut next_hash = reference_hash;
+        let mut num_requests = 0;
         for _ in 0..max_block_requests {
-            match chain.chain_store().get_next_block_hash(&next_hash) {
-                Ok(hash) => next_hash = hash,
+            next_hash = match chain.chain_store().get_next_block_hash(&next_hash) {
+                Ok(hash) => hash,
                 Err(e) => match e {
-                    near_chain::Error::DBNotFoundErr(_) => break,
+                    near_chain::Error::DBNotFoundErr(_) => {
+                        debug!(
+                            target: "sync",
+                            block_hash = ?next_hash,
+                            "next block hash is not found"
+                        );
+                        break;
+                    }
                     _ => return Err(e),
                 },
+            };
+            if let Err(err) = check_known(chain, &next_hash)? {
+                debug!(
+                    target: "sync",
+                    block_hash = ?next_hash,
+                    ?err,
+                    "block is known"
```

Comment:
Huh, the API is funky. Does it really return `Err(e)` when the block is already known? Feels like method should be called `ensure_unknown` or something…

Reply:
yes, 100% agree, renamed and moved as member function to `Chain` in bf27be626